mitodl/micromasters
# Fixed exam authorization command and refactoring

#### What are the relevant tickets?
- fixes https://github.com/mitodl/micromasters/issues/2445
- Fixed ripples of #2331

#### What's this PR do?
- Refactored the code of exam authorization command
- Added exceptions and logging to see behaviour of code
- Added new tests for authorization code.

### How to test:
- In .env add `FEATURE_FINAL_GRADE_ALGORITHM=v1`
- Set program.exam_series_code
- Set course.exam_module
- Pay for course
- Create final grade http://192.168.99.100:8079/admin/grades/finalgrade/
- Run command ```docker-compose run web ./manage.py authorization_user_exam --username=(value) --program-id=(value)```
- Check weather ( ExamAuthoriztion) object created 
  - ``docker-compose run web ./manage.py shell``
  - ``from exams.models import ExamAuthorization``
  -  ``ExamAuthorization.objects.all()``

#### To create dummy data:
```docker-compose run web ./manage.py shell```

```shell
from courses.models import CourseRun
from dashboard.models import CachedEnrollment
from dashboard.factories import CachedEnrollmentFactory
from ecommerce.factories import (OrderFactory, LineFactory)
from django.contrib.auth.models import User
from exams.models import ExamAuthorization

user = User.objects.get(username='amirq')
# use same  `edx_course_key` for final grade entry
course_run = CourseRun.objects.get(edx_course_key='course-v1:MITx+Digital+Learning+100+Jan_2016')
CachedEnrollmentFactory.create(user=user, course_run=course_run)
# enter  `total_price_paid` according to income threshold
order = OrderFactory.create(user=user, status='fulfilled', total_price_paid=1200)
LineFactory.create(order=order, course_key=course_run.edx_course_key)
ExamAuthorization.objects.all()
```

@pdpinch @giocalitri
